backend/vapid.js

```diff
@@ -0,0 +1,95 @@
+import sbp from '@sbp/sbp'
+
+let vapidPublicKey: string
+let vapidPrivateKey: Object
+
+// The Voluntary Application Server Identification (VAPID) email field is "a
+// stable identity for the application server" that "can be used by a push
+// service to establish behavioral expectations for an application server"
+// RFC 8292
+if (!process.env.VAPID_EMAIL) {
+  console.warn('Missing VAPID identification. Please set VAPID_EMAIL to a value like "mailto:some@example".')
+}
+const vapid = { VAPID_EMAIL: process.env.VAPID_EMAIL || 'mailto:test@example.com' }
+
+export const initVapid = async () => {
+  const vapidKeyPair = await sbp('chelonia/db/get', '_private_immutable_vapid_key').then(async (vapidKeyPair: string): Promise<[Object, string]> => {
+    if (!vapidKeyPair) {
+      // Generate a new ECDSA key pair
+      const keyPair = await crypto.subtle.generateKey(
+        {
+          name: 'ECDSA',
+          namedCurve: 'P-256' // Use P-256 curve
+        },
+        true, // Whether the key is extractable
+        ['sign', 'verify'] // Usages
+      )
+
+      // Export the private key
+      const serializedKeyPair = await Promise.all([
+        crypto.subtle.exportKey('jwk', keyPair.privateKey),
+        crypto.subtle.exportKey('raw', keyPair.publicKey).then((key) =>
+          // $FlowFixMe[incompatible-call]
+          Buffer.from(key).toString('base64url')
+        )
+      ])
+
+      return sbp('chelonia/db/set', '_private_immutable_vapid_key', JSON.stringify(serializedKeyPair)).then(() => {
+        return [keyPair.privateKey, serializedKeyPair[1]]
+      })
+    }
+
+    const serializedKeyPair = JSON.parse(vapidKeyPair)
+    return [
+      await crypto.subtle.importKey(
+        'jwk',
+        serializedKeyPair[0],
+        { name: 'ECDSA', namedCurve: 'P-256' },
+        false,
+        ['sign']
+      ),
+      serializedKeyPair[1]
+    ]
+  })
+
+  vapidPrivateKey = vapidKeyPair[0]
+  vapidPublicKey = vapidKeyPair[1]
+}
+
+const generateJwt = async (endpoint: URL): Promise<string> => {
+  const now = Date.now() / 1e3 | 0
+
+  const audience = endpoint.origin
+
+  const header = Buffer.from(JSON.stringify(
+    Object.fromEntries([['typ', 'JWT'], ['alg', 'ES256']])
+    // $FlowFixMe[incompatible-call]
+  )).toString('base64url')
+  const body = Buffer.from(JSON.stringify(
+    Object.fromEntries([
+      ['aud', audience],
+      ['exp', now + 90],
+      ['iat', now],
+      ['nbf', now - 90],
```

Comment:
Claude thinks these `90` values are "quite short"... I know nothing about this so can only pass along the comment as a question — is Claude right?

Reply:
Claude is right. That's 90s. Since we're pushing to a remote server, there's no reason to make the token long lived. Originally it was 60s, but for some reason Google seemed to reject them, so I increased them to 90s, which seemed to work.

Reply:
If you want, I can change it to some other value, but keep in mind that the token is intended for immediate use (for reference, `nbf` is 'not before' and `exp` is 'not after')

Reply:
I have no preference, just asking